pkg/solana/txm/txm_internal_test.go

```diff
@@ -1090,3 +1177,390 @@ func addSigAndLimitToTx(t *testing.T, keystore SimpleKeystore, pubkey solana.Pub
+				func(_ context.Context, sigs []solana.Signature) ([]*rpc.SignatureStatusesResult, error) {
+					var out []*rpc.SignatureStatusesResult
+					for _, sig := range sigs {
+						getStatus, exists := statuses[sig]
+						if !exists {
+							out = append(out, nil)
+						} else {
+							out = append(out, getStatus())
+						}
+					}
+					return out, nil
+				},
+			).Maybe()
+		}
+
+		mkey := keyMocks.NewSimpleKeystore(t)
+		mkey.On("Sign", mock.Anything, mock.Anything, mock.Anything).Return([]byte{}, nil)
+
+		loader := utils.NewLazyLoad(func() (client.ReaderWriter, error) { return mc, nil })
+		txm := NewTxm(id, loader, nil, cfg, mkey, lggr)
+		require.NoError(t, txm.Start(ctx))
+		t.Cleanup(func() { require.NoError(t, txm.Close()) })
+
+		return txm, mc, mkey
+	}
+
+	t.Run("WithRebroadcast", func(t *testing.T) {
+		txExpirationRebroadcast := true
+		statuses := map[solana.Signature]func() *rpc.SignatureStatusesResult{}
+
+		// Mock SlotHeight to return a value greater than 0
+		slotHeightFunc := func() (uint64, error) {
+			return uint64(1500), nil
+		}
+
+		callCount := 0
+		latestBlockhashFunc := func() (*rpc.GetLatestBlockhashResult, error) {
+			defer func() { callCount++ }()
+			if callCount < 1 {
+				// To force rebroadcast, first call needs to be smaller than slotHeight
+				return &rpc.GetLatestBlockhashResult{
+					Value: &rpc.LatestBlockhashResult{
+						LastValidBlockHeight: uint64(1000),
+					},
+				}, nil
+			}
+			// following rebroadcast call will go through because lastValidBlockHeight is bigger than slotHeight
+			return &rpc.GetLatestBlockhashResult{
+				Value: &rpc.LatestBlockhashResult{
+					LastValidBlockHeight: uint64(2000),
+				},
+			}, nil
+		}
+
+		sig1 := randomSignature(t)
+		sendTxFunc := func() (solana.Signature, error) {
+			return sig1, nil
+		}
+
+		nowTs := time.Now()
+		sigStatusCallCount := 0
+		var wg sync.WaitGroup
+		wg.Add(1)
+		statuses[sig1] = func() *rpc.SignatureStatusesResult {
+			// First transaction should be rebroadcasted.
+			if time.Since(nowTs) < cfg.TxConfirmTimeout()-2*time.Second {
+				return nil
+			} else {
+				// Second transaction should reach finalization.
+				sigStatusCallCount++
+				if sigStatusCallCount == 1 {
+					return &rpc.SignatureStatusesResult{
+						ConfirmationStatus: rpc.ConfirmationStatusProcessed,
+					}
+				} else if sigStatusCallCount == 2 {
+					return &rpc.SignatureStatusesResult{
+						ConfirmationStatus: rpc.ConfirmationStatusConfirmed,
+					}
+				} else {
+					wg.Done()
+					return &rpc.SignatureStatusesResult{
+						ConfirmationStatus: rpc.ConfirmationStatusFinalized,
+					}
+				}
+			}
+		}
+
+		txm, _, mkey := setupTxmTest(txExpirationRebroadcast, latestBlockhashFunc, slotHeightFunc, sendTxFunc, statuses)
+
+		tx, _ := getTx(t, 0, mkey)
+		txID := "test-rebroadcast"
+		assert.NoError(t, txm.Enqueue(ctx, t.Name(), tx, &txID))
+		wg.Wait()
+		time.Sleep(2 * time.Second) // Sleep to allow for rebroadcasting
```

Comment:
You don't have to use this if 2 second sleep is deterministic so rebroadcast always happen in that time. If not, maybe it's safer to use something like `waitFor(t, waitDuration, txm, prom, empty)` which we use in other tests above to wait till all tx have gone through.